### PR TITLE
Better breakpoint support when clicking

### LINKED ISSUE #2814: Better breakpoint support when clicking (works fine with F9)
It is not possible to fine put fine grained breakpoints by clicking but only with F9.

#### Repro steps

Provide the steps required to reproduce the problem

1. Suppose we have the below code:

```
[<EntryPoint>]
let main argv = 
    let x = 1
    
    let z =
        match x with
        | 1 ->  x
        | _ ->  x + 1
    
    printfn "%A" (z)
    0 // return an integer exit code
```

2. We try to click to left hand side of visual studio to line starting `| 1 -> x` . Visual studio puts the breakpoint to the match statement. 

3. However if we click to `x` then hit F9 then a proper breakpoint inserted.

#### Expected behavior

The breakpoint should be put the relevant line.

#### Actual behavior

The breakpoint is put to the nearest match statement.

#### Known workarounds

Manual mouse click and hit F9 puts the breakpoint to the desired location.
#### Related information 

VS 2017 RTM (all latest updates installed), F# tools nightly.

### PR DESCRIPTION
**Pull request by forki**
fixes #2814

### CODE REVIEW

**Comment by forki** [comment]
I also made it so that it does not set breakpoints if the line question is empty. Just as C# behaves,

File: src/fsharp/vs/ServiceUntypedParse.fs [hunk 1]
```diff
@@ -351,12 +351,24 @@ type FSharpParseFileResults(errors : FSharpErrorInfo[], input : Ast.ParsedInput
...
-                | [] ->
-                    match locations |> List.filter (fun m -> rangeBeforePos m pos |> not) with
-                    | [] -> Seq.tryHead locations
-                    | locationsAfterPos -> Seq.tryHead locationsAfterPos
-                | coveringLocations -> Seq.tryLast coveringLocations)
+                if pos.Column = 0 then
+                    // we have a breakpoint that was set with mouse at line start
+                    match locations |> List.filter (fun m -> m.StartLine = m.EndLine && pos.Line = m.StartLine) with
+                    | [] ->
+                        match locations |> List.filter (fun m -> rangeContainsPos m pos) with
```
Comment (KevinRansom) [inline src/fsharp/vs/ServiceUntypedParse.fs]: Is this the same code as the else clause ... perhaps a minor refactor?

Comment (forki) [inline src/fsharp/vs/ServiceUntypedParse.fs]: I assume we change that soon. I don't want to make it DRY yet. Can you please still merge it?

Comment (KevinRansom) [inline src/fsharp/vs/ServiceUntypedParse.fs]: For my education why do you assume it will change soon?

Comment (forki) [inline src/fsharp/vs/ServiceUntypedParse.fs]: The whole thing feels wrong. I can't really describe it but I think there is much more to do than just making this dry. That said: the fix actually solves a important bug. So...

**Comment by KevinRansom** [comment]
Sure.

Thanks for taking care of this.

File: src/fsharp/vs/ServiceUntypedParse.fs (hunk 1 above)
Comment (saul) [inline src/fsharp/vs/ServiceUntypedParse.fs]: Would it really be difficult to abstract this into a function? It's such a small change and it makes it clear that the behaviour of both should be identical (for now)

**Comment by KevinRansom** [comment]
@saul It's merged, but I would be happy to accept a PR :-)

**Comment by saul** [comment]
D'oh - I should learn to read!

**Comment by forki** [comment]
@saul this code duplication is basically an invitation for others to reflect on this and give good input. Making it DRY is easy, but I'm looking for real feedback on "what do we really want to do here?".

**Comment by KevinRansom** [comment]
@forki perhaps you can start a conversation with something a bit more substantial than "feels" wrong.  

- Why does it feel wrong?  
- What should it achieve that it doesn't etcetera?

Sure looking at the repetition is low hanging review fruit?  But it is also not a thing we typically expect when performing a code review.  So I think a little guidance into what you think is wrong, may allow us to think about the problem little more precisely.

Kevin

**Comment by forki** [comment]
Kevin you are absolutely right and I apologize.
But here it's really just a strange feeling. I don't know how to express
it.
Sorry

Am 07.04.2017 10:37 nachm. schrieb "Kevin Ransom (msft)" <
notifications@github.com>:

**Pull request merged**